File: apps/hermes-desktop-tauri/src/types/task.ts
export type TaskStatus =
  | 'pending'
  | 'running'
  | 'needs_approval'
  | 'done'
  | 'failed'
  | 'cancelled'
  | 'scheduled'
  | 'paused'

export type EventKind =
  | 'instruction'
  | 'plan'
  | 'thinking'
  | 'tool_call'
  | 'tool_result'
  | 'subagent_spawn'
  | 'message'
  | 'artifact'
  | 'approval_request'
  | 'approval_response'
  | 'checkpoint'
  | 'error'
  | 'system'

export type TocIcon =
  | 'message'
  | 'plan'
  | 'thinking'
  | 'tool'
  | 'artifact'
  | 'approval'
  | 'checkpoint'
  | 'error'
  | 'fork'

export interface TokenUsage {
  input_tokens: number
  output_tokens: number
  cost_usd_cents: number
}

export interface AgentPersona {
  vertical_id?: string | null
  system_prompt: string
  model_id?: string | null
  provider_id?: string | null
}

export interface CronSchedule {
  expr: string
  timezone: string
  next_run?: string | null
  last_run?: string | null
  enabled: boolean
}

export interface Task {
  id: string
  owner_user_id: string
  primary_device_id: string
  title: string
  vertical?: string | null
  status: TaskStatus
  parent_task_id?: string | null
  persona_stack: AgentPersona[]
  schedule?: CronSchedule | null
  created_at: string
  updated_at: string
}

export type TaskActor =
  | { type: 'user'; user_id: string; device_id: string }
  | { type: 'agent'; model_id: string; provider_id: string }
  | { type: 'tool'; tool_name: string }
  | { type: 'system' }

export interface TaskEvent {
  id: string
  task_id: string
  parent_event_id?: string | null
  kind: EventKind
  actor: TaskActor
  payload: Record<string, unknown>
  collapsed_by_default: boolean
  streaming: boolean
  created_at: string
  duration_ms?: number | null
  cost_tokens?: TokenUsage | null
  turn_id?: string | null
  toc_label?: string | null
  toc_icon?: TocIcon | null
  anchor_slug: string
}

export interface TaskTurn {
  id: string
  task_id: string
  instruction_event_id: string
  label: string
  started_at: string
  ended_at?: string | null
  status: 'running' | 'done' | 'cancelled' | 'failed'
  artifact_count: number
  approval_count: number
  error_count: number
  cost_tokens: TokenUsage
  sub_task_ids: string[]
}

export interface VerticalMeta {
  id: string
  display_name_key: string
  description_key: string
  icon: string
  category: string
  order: number
  task_category?: string | null
}

export interface TaskListParams {
  owner_user_id?: string
  status?: TaskStatus
  vertical?: string
  cursor?: string
  limit?: number
}

export interface TaskListResponse {
  tasks: Task[]
  next_cursor?: string | null
}

export interface CreateTaskRequest {
  title: string
  vertical?: string
  instruction?: string
  owner_user_id?: string
  device_id?: string
}

export interface CreateTaskResponse {
  task: Task
  event: TaskEvent
}

export interface TaskEventsResponse {
  events: TaskEvent[]
}

export interface TaskTurnsResponse {
  turns: TaskTurn[]
}

export interface VerticalListResponse {
  verticals: VerticalMeta[]
}

export const TASK_STATUS_GROUPS: { key: string; statuses: TaskStatus[] }[] = [
  { key: 'running', statuses: ['running', 'pending'] },
  { key: 'approval', statuses: ['needs_approval'] },
  { key: 'scheduled', statuses: ['scheduled', 'paused'] },
  { key: 'done', statuses: ['done', 'failed', 'cancelled'] }
]

File: crates/hermes-tasks/src/types/task.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::ids::{DeviceId, TaskId, UserId, VerticalId};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TaskStatus {
    Pending,
    Running,
    NeedsApproval,
    Done,
    Failed,
    Cancelled,
    Scheduled,
    Paused,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentPersona {
    pub vertical_id: Option<VerticalId>,
    pub system_prompt: String,
    pub model_id: Option<String>,
    pub provider_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CronSchedule {
    pub expr: String,
    pub timezone: String,
    pub next_run: Option<DateTime<Utc>>,
    pub last_run: Option<DateTime<Utc>>,
    pub enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub id: TaskId,
    pub owner_user_id: UserId,
    pub primary_device_id: DeviceId,
    pub title: String,
    pub vertical: Option<VerticalId>,
    pub status: TaskStatus,
    pub parent_task_id: Option<TaskId>,
    pub persona_stack: Vec<AgentPersona>,
    pub schedule: Option<CronSchedule>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Task {
    pub fn new(
        owner_user_id: UserId,
        primary_device_id: DeviceId,
        title: impl Into<String>,
        vertical: Option<VerticalId>,
    ) -> Self {
        let now = Utc::now();
        Self {
            id: TaskId::new(),
            owner_user_id,
            primary_device_id,
            title: title.into(),
            vertical,
            status: TaskStatus::Pending,
            parent_task_id: None,
            persona_stack: Vec::new(),
            schedule: None,
            created_at: now,
            updated_at: now,
        }
    }

    pub fn with_parent(mut self, parent_task_id: TaskId) -> Self {
        self.parent_task_id = Some(parent_task_id);
        self
    }

    pub fn with_persona(mut self, persona: AgentPersona) -> Self {
        self.persona_stack.push(persona);
        self
    }

    pub fn with_schedule(mut self, schedule: CronSchedule) -> Self {
        self.schedule = Some(schedule);
        self.status = TaskStatus::Scheduled;
        self
    }
}
